lightning/src/offers/merkle.rs

```diff
@@ -0,0 +1,178 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Tagged hashes for use in signature calculation and verification.
+
+use bitcoin::hashes::{Hash, HashEngine, sha256};
+use bitcoin::secp256k1::{Message, PublicKey, Secp256k1, self};
+use bitcoin::secp256k1::schnorr::Signature;
+use crate::io;
+use crate::util::ser::{BigSize, Readable};
+
+use crate::prelude::*;
+
+/// Valid type range for signature TLV records.
+const SIGNATURE_TYPES: core::ops::RangeInclusive<u64> = 240..=1000;
+
+tlv_stream!(SignatureTlvStream, SignatureTlvStreamRef, SIGNATURE_TYPES, {
+	(240, signature: Signature),
+});
+
+pub(super) fn sign_message<F>(
+	sign: F, tag: &str, bytes: &[u8], pubkey: PublicKey,
+) -> Result<Signature, secp256k1::Error>
+where
+	F: FnOnce(&Message) -> Signature
+{
+	let digest = message_digest(tag, bytes);
+	let signature = sign(&digest);
+
+	let pubkey = pubkey.into();
+	let secp_ctx = Secp256k1::verification_only();
+	secp_ctx.verify_schnorr(&signature, &digest, &pubkey)?;
+
+	Ok(signature)
+}
+
+/// Verifies the signature with a pubkey over the given bytes using a tagged hash as the message
+/// digest.
+pub(super) fn verify_signature(
+	signature: &Signature, tag: &str, bytes: &[u8], pubkey: PublicKey,
+) -> Result<(), secp256k1::Error> {
+	let digest = message_digest(tag, bytes);
+	let pubkey = pubkey.into();
+	let secp_ctx = Secp256k1::verification_only();
+	secp_ctx.verify_schnorr(signature, &digest, &pubkey)
+}
+
+fn message_digest(tag: &str, bytes: &[u8]) -> Message {
+	let tag = sha256::Hash::hash(tag.as_bytes());
+	let merkle_root = root_hash(bytes);
+	Message::from_slice(&tagged_hash(tag, merkle_root)).unwrap()
+}
+
+/// Computes a merkle root hash for the given data, which must be a well-formed TLV stream
+/// containing at least one TLV record.
+fn root_hash(data: &[u8]) -> sha256::Hash {
+	let mut tlv_stream = TlvStream::new(&data[..]).peekable();
+	let nonce_tag = tagged_hash_engine(sha256::Hash::from_engine({
+		let mut engine = sha256::Hash::engine();
+		engine.input("LnNonce".as_bytes());
+		engine.input(tlv_stream.peek().unwrap().type_bytes);
```

Comment:
> Two questions -
> a) is this unwrap definitely going to always remain safe? Should we instead just make it an if statement to make it trivial?

Hmm... it should always be safe because `invreq_metadata` is type 0 and MUST be set. But it appears I'm not checking that. 😞  

> b) wait, what? So now we add a nonce to each leaf that is only the first TLV record in the offer/invoice/etc? That doesn't seem sufficient, no? The nonce needs to be unpredictable such that an attacker can't reasonably guess it, AFAIU, is the first TLV record always unpredictable?

It should always be `invreq_metadata`, so the BOLT says: "It's also the first entry (if present), ensuring an unpredictable nonce for hashing." Not sure why "if present" is mentioned, though.

Reply:
Pushed fixups to make it required.

Reply:
Yeah, at least two utilities:
- one for creating a stateless offer which would populate the `offer_metadata` such that it can be checked when handling an `invoice_request` message.
- another for constructing an `invoice_request` which would populate `invreq_metadata` such that it can be checked when handling an `invoice` message.